xbmc/addons/kodi-addon-dev-kit/include/kodi/addon-instance/PVR.h

```diff
@@ -0,0 +1,3076 @@
+                                                 const PVR_RECORDING* recording,
+                                                 int64_t* size)
+  {
+    return static_cast<CInstancePVRClient*>(instance->toAddon->addonInstance)
+        ->GetRecordingSize(recording, *size);
+  }
+
+  inline static PVR_ERROR ADDON_GetRecordingStreamProperties(const AddonInstance_PVR* instance,
+                                                             const PVR_RECORDING* recording,
+                                                             PVR_NAMED_VALUE* properties,
+                                                             unsigned int* propertiesCount)
+  {
+    std::vector<PVRStreamProperty> propertiesList;
+    PVR_ERROR error = static_cast<CInstancePVRClient*>(instance->toAddon->addonInstance)
+                          ->GetRecordingStreamProperties(recording, propertiesList);
+    *propertiesCount = 0;
+    for (auto iter : propertiesList)
+    {
+      strncpy(properties[*propertiesCount].strName, iter.GetName().c_str(),
+              sizeof(properties[*propertiesCount].strName) - 1);
+      strncpy(properties[*propertiesCount].strValue, iter.GetValue().c_str(),
+              sizeof(properties[*propertiesCount].strValue) - 1);
+      ++*propertiesCount;
+      if (*propertiesCount > STREAM_MAX_PROPERTY_COUNT)
+        break;
+    }
+    return error;
+  }
+
+  inline static PVR_ERROR ADDON_CallRecordingMenuHook(const AddonInstance_PVR* instance,
+                                                      const PVR_MENUHOOK* menuhook,
+                                                      const PVR_RECORDING* recording)
+  {
+    return static_cast<CInstancePVRClient*>(instance->toAddon->addonInstance)
+        ->CallRecordingMenuHook(menuhook, recording);
+  }
+
+  //--==----==----==----==----==----==----==----==----==----==----==----==----==
+
+
+  inline static PVR_ERROR ADDON_GetTimerTypes(const AddonInstance_PVR* instance,
+                                              PVR_TIMER_TYPE* types,
+                                              int* typesCount)
+  {
+    std::vector<PVRTimerType> timerTypes;
+    PVR_ERROR error = static_cast<CInstancePVRClient*>(instance->toAddon->addonInstance)
+                          ->GetTimerTypes(timerTypes);
+    *typesCount = 0;
+    for (auto iter : timerTypes)
+    {
+      types[*typesCount] = *iter;
+      ++*typesCount;
+      if (*typesCount >= PVR_ADDON_TIMERTYPE_ARRAY_SIZE)
+        break;
+    }
+    return error;
+  }
+
+  inline static int ADDON_GetTimersAmount(const AddonInstance_PVR* instance)
+  {
+    return static_cast<CInstancePVRClient*>(instance->toAddon->addonInstance)->GetTimersAmount();
+  }
+
+  inline static PVR_ERROR ADDON_GetTimers(const AddonInstance_PVR* instance, ADDON_HANDLE handle)
+  {
+    std::vector<PVRTimer> timers;
+    PVR_ERROR error =
+        static_cast<CInstancePVRClient*>(instance->toAddon->addonInstance)->GetTimers(timers);
+    for (auto iter : timers)
```

Comment:
in case of error I would not call the callback

Reply:
and you should use `const auto& iter`, here and at every other place where you iterate a vector to call a callback for every element.

Reply:
and last, range base for loop does not denotify an iterator, its actually the dereferenced iterator, a `timer`in this case. Applies also to several places in this file.